src/layers/attention/self_attention.rs

```diff
@@ -387,38 +387,28 @@ impl Attention for SelfAttention {
             key = key_rot;
         }
 
-        let (key, value, attention_mask) = match cache {
-            Some(cache) => (
-                // Shape is: [batch_size, n_heads, key_len, head_width], so
-                // concat along dimension 2.
-                Tensor::cat(&[&cache.key, &key], 2).context(ConcatKVCacheSnafu)?,
-                Tensor::cat(&[&cache.value, &value], 2).context(ConcatKVCacheSnafu)?,
-                cache
-                    .attention_mask
-                    .extend(attention_mask)
-                    .context(AttentionMaskSnafu)?,
-            ),
-            None => (key, value, attention_mask.clone()),
-        };
+        cache.update(&key, &value)?;
+        let key = cache.key().unwrap_or(&key);
+        let value = cache.value().unwrap_or(&value);
```

Comment:
Looks like we don't cache the attention mask anymore? Doesn't look like it's intentional?

Reply:
I should have put this in the commit message. It's intentional, because I am not sure where it should go. Before we were doing a lot of unnecessary allocations because we stored an attention mask for every layer.

But if you store it only once in the cache, who is responsible for updating the mask? Not a particular layer, because then other layers need to know that they are not the first layer. I guess it could be done in the decoder, but then the decoder and decoder layer interfaces would diverge - the decoder would only expect an attention mask for the new tokens and the layers for all tokens.

Reply:
Updated the commit message with more information why the mask is removed from the cache for the time being.

Reply:
Makes sense.